Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp

```diff
@@ -0,0 +1,674 @@
+
+      unsigned int aid1, aid2, aid3, aid4;
+      unsigned int bid2, bid1, bid3, id1, id2;
+
+      BIT_SET doneBonds(nb);
+
+      if (useExpTorsions) {
+        // we set the torsion angles with experimental data
+        const ExpTorsionAngleCollection *params = ExpTorsionAngleCollection::getParams();
+
+        // loop over patterns
+        for (ExpTorsionAngleCollection::ParamsVect::const_iterator it = params->begin(); it != params->end(); ++it) {
+          std::vector<MatchVectType> matches;
+          SubstructMatch(mol, *(it->dp_pattern.get()), matches, false, true);
+
+          // loop over matches
+          for(std::vector<MatchVectType>::const_iterator matchIt = matches.begin();
+            matchIt != matches.end(); ++matchIt) {
+            // get bond indices
+            aid1 = (*matchIt)[it->idx[0]].second;
+            aid2 = (*matchIt)[it->idx[1]].second;
+            aid3 = (*matchIt)[it->idx[2]].second;
+            aid4 = (*matchIt)[it->idx[3]].second;
+            // FIX: check if bond is NULL
+            bid2 = mol.getBondBetweenAtoms(aid2, aid3)->getIdx();
+            if (!doneBonds[bid2]) {
+              doneBonds[bid2] = 1;
+              std::vector<int> atoms(4);
+              atoms[0] = aid1; atoms[1] = aid2; atoms[2] = aid3; atoms[3] = aid4;
+              expTorsionAtoms.push_back(atoms);
+              expTorsionAngles.push_back(std::make_pair(it->signs, it->V));
+              if (verbose) {
+                std::cout << it->smarts << ": " << aid1 << " " << aid2 << " " << aid3 << " " << aid4 << ", (";
+                for (unsigned int i = 0; i < it->V.size()-1; ++i) {
+                  std::cout << it->V[i] << ", ";
+                }
+                std::cout << it->V[it->V.size()-1] << ") " << std::endl;
+              }
+            } // if not donePaths
+          } // end loop over matches
+
+        } // end loop over patterns
+      }
+
+      // apply basic knowledge such as flat aromatic rings, other sp2-centers,
+      // straight triple bonds, etc.
+      if (useBasicKnowledge) {
+
+        BIT_SET doneAtoms(na);
+        ROMol::ADJ_ITER nbrIdx;
+        ROMol::ADJ_ITER endNbrs;
+
+        // inversion terms (improper torsions / out-of-plane bends / inversion)
+        // loop over atoms
+        for (int aid2 = 0; aid2 < na; ++aid2) {
+          if (!(doneAtoms[aid2])) {
+            std::vector<int> atoms(4, -1);
+            atoms[1] = aid2;
+            const Atom *atom2 = mol.getAtomWithIdx(atoms[1]);
+            int at2AtomicNum = atom2->getAtomicNum();
+
+            // if atom is a N,O or C and SP2-hybridized
+            if (((at2AtomicNum == 6) || (at2AtomicNum == 7) || (at2AtomicNum == 8))
+                && (atom2->getHybridization() == Atom::SP2)) {
+              // get neighbors
+              boost::tie(nbrIdx, endNbrs) = mol.getAtomNeighbors(atom2);
+              // check if enough neighbours
+              if (std::distance(nbrIdx,endNbrs) != 3) {
```

Comment:
probably more efficient to just call `mol.getAtomDegree(atom2)` before line 598 instead of doing this check here.